File: src/main/java/uk/gov/justice/digital/hmpps/oauth2server/security/MaintainUserCheck.java
package uk.gov.justice.digital.hmpps.oauth2server.security;

import com.google.common.collect.Sets;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.stereotype.Service;
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.User;
import uk.gov.justice.digital.hmpps.oauth2server.auth.repository.UserRepository;

import java.util.Collection;

@Service
public class MaintainUserCheck {

    private final UserRepository userRepository;

    public MaintainUserCheck(
            final UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    public static boolean canMaintainAuthUsers(final Collection<? extends GrantedAuthority> authorities) {
        return authorities.stream().map(GrantedAuthority::getAuthority).anyMatch("ROLE_MAINTAIN_OAUTH_USERS"::equals);
    }

    public void ensureUserLoggedInUserRelationship(final String loggedInUser, final Collection<? extends GrantedAuthority> authorities, final User user) throws AuthUserGroupRelationshipException {
        // if they have maintain privileges then all good
        if (canMaintainAuthUsers(authorities)) {
            return;
        }
        // otherwise group managers must have a group in common for maintenance
        final var loggedInUserEmail = userRepository.findByUsernameAndMasterIsTrue(loggedInUser).orElseThrow();
        if (Sets.intersection(loggedInUserEmail.getGroups(), user.getGroups()).isEmpty()) {
            // no group in common, so disallow
            throw new AuthUserGroupRelationshipException(user.getName(), "User not with your groups");
        }
    }

    public static class AuthUserGroupRelationshipException extends Exception {
        private final String username;
        private final String errorCode;

        public AuthUserGroupRelationshipException(final String username, final String errorCode) {
            super(String.format("Unable to maintain user: %s with reason: %s", username, errorCode));

            this.username = username;
            this.errorCode = errorCode;
        }

        public String getUsername() {
            return this.username;
        }

        public String getErrorCode() {
            return this.errorCode;
        }
    }
}

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/security/MaintainUserCheck.kt
package uk.gov.justice.digital.hmpps.oauth2server.security

import com.google.common.collect.Sets
import org.springframework.security.core.GrantedAuthority
import org.springframework.stereotype.Service
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.User
import uk.gov.justice.digital.hmpps.oauth2server.auth.repository.UserRepository

@Service
class MaintainUserCheck(
  private val userRepository: UserRepository
) {

  companion object {
    fun canMaintainAuthUsers(authorities: Collection<GrantedAuthority>): Boolean {
      return authorities.stream().map { obj: GrantedAuthority -> obj.authority }
        .anyMatch { "ROLE_MAINTAIN_OAUTH_USERS" == it }
    }
  }

  @Throws(AuthUserGroupRelationshipException::class)
  fun ensureUserLoggedInUserRelationship(loggedInUser: String, authorities: Collection<GrantedAuthority>, user: User) {
    // if they have maintain privileges then all good
    if (canMaintainAuthUsers(authorities)) {
      return
    }
    // otherwise group managers must have a group in common for maintenance
    val loggedInUserEmail = userRepository.findByUsernameAndMasterIsTrue(loggedInUser).orElseThrow()
    if (Sets.intersection(loggedInUserEmail.groups, user.groups).isEmpty()) {
      // no group in common, so disallow
      throw AuthUserGroupRelationshipException(user.name, "User not with your groups")
    }
  }

  class AuthUserGroupRelationshipException(val username: String, val errorCode: String) :
    Exception(String.format("Unable to maintain user: %s with reason: %s", username, errorCode))
}
